Decoupled cMonster and path recalc logic, re-implemented recalc.

SafwatHalaby opened the pull request:
~~Do not merge yet. This needs review. This is in pretty good shape, but not yet complete.~~

Fixes #2536

Known issues:
 - **Fixed**, Mobs sometimes do not detect that they've reached their final destination. The result is that passive mobs sometimes just sit still. 
 - **Fixed**, Needs to conform to the lua style check script.
 - **Fixed**, It appears I've reverted some commits made by others,  need to fix that.
 - **Fixed**, Need to review and update any out-of-date AI-related doxy docs in the headers.

SafwatHalaby commented on src/Mobs/Monster.cpp:
```diff
@@ -123,127 +120,20 @@ void cMonster::SpawnOn(cClientHandle & a_Client)
 		{
 			if (
-				(IsOnGround() && (GetSpeedX() == 0.0f) && (GetSpeedY() == 0.0f)) ||
-				(IsSwimming() && (m_GiveUpCounter < 15))
```
I don't remember touching these two lines.

madmaxoft replied:
This looks like the fix to avoid taking fall damage while swimming in water

SafwatHalaby replied:
TODO investigate this

SafwatHalaby replied:
Monster.cpp no longer has a `GiveUpCounter`, this is why I edited this line. It breaks jumping in water, but AI water handling is pretty broken altogether, so this isn't an issue. We'll get back to this when water handling is fixed.

SafwatHalaby commented on src/Mobs/Monster.h:
```diff
@@ -117,8 +112,8 @@ class cMonster :
 	// tolua_begin
 	bool IsBaby (void) const { return m_Age < 0; }
-	int GetAge (void) const { return m_Age; }
-	void SetAge(int a_Age)  { m_Age = a_Age; }
```
Umm, I think I've overridden someone's commit/s somehow. Did I copy-paste a whole file at some point?

SafwatHalaby commented:
Clang error:
```
     conversion loses integer precision: 'unsigned long' to 'int'

      [-Werror,-Wshorten-64-to-32]

                return m_PathPoints.size() - m_CurrentPoint;

                ~~~~~~ ~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~
```

SafwatHalaby commented on src/Mobs/PathFinder.cpp:
```diff
@@ -0,0 +1,237 @@
+}
+
+// TODO fix monster heads
+// TODO why does PATH_NOT_FOUND even occur?
```
Irrelevant comments, already fixed in this commit

tigerw commented:
D'you know what? Why don't you just merge yourself after you've finished your own review? :P

NiLSPACE commented on src/Mobs/Monster.cpp:
```diff
@@ -123,127 +120,20 @@ void cMonster::SpawnOn(cClientHandle & a_Client)
-	}
-
-	if (ReachedFinalDestination())
+	if ((m_NextWayPointPosition - GetPosition()).Length() < GetWidth()/2)
```
Space around the `/`

Also, add extra parentheses to conditions. `(GetWidth() / 2)`

madmaxoft replied:
Nah, the extra parenthesis is not really needed.

SphinxC0re replied:
Why not? The convention says that it needs to be like this

madmaxoft replied:
The convention says "parenthesis around comparisons", not "parenthesis around comparison operands"
Similarly, we write `if (a > b + 2)` instead of `if (a > (b + 2))`

SphinxC0re replied:
But isn't that bad? It looks awful

SafwatHalaby replied:
The lua script missed this one.

NiLSPACE commented on src/Mobs/PathFinder.cpp:
```diff
@@ -0,0 +1,237 @@
+		ResetPathFinding(a_Chunk);
+	}
+
+	switch (m_Path.Step(a_Chunk))
```
I believe some people were not happy with the naming of `Step`. Perhaps this is the time to get a proper name for it?

SphinxC0re replied:
What about **find* instead?

SafwatHalaby replied:
Suggestions are welcome. I can't find a better name.

SphinxC0re replied:
What does it do?

SafwatHalaby replied:
To prevent blocking for hard path calculations, the calculation is split into ticks. Each tick does a portion of the calculation. However, the word `tick` can be confused with Minecraft ticks, so I called it `step`.

Each call calculates a small part of the path, until the calculation is done.

SphinxC0re replied:
```CalculationStep``` ?

SafwatHalaby replied:
Is that less vague than step?

SphinxC0re replied:
I guess so

NiLSPACE commented on src/Mobs/PathFinder.cpp:
```diff
@@ -0,0 +1,237 @@
+
+	return true;
+}
+
```
Enters after methods/functions.

SphinxC0re replied:
"Empty lines"

SafwatHalaby replied:
Fixed.

SphinxC0re commented on src/Mobs/PathFinder.cpp:
```diff
@@ -0,0 +1,237 @@
+#include "Globals.h"
+#include "PathFinder.h"
+#include "../Chunk.h"
+
```
add 5 empty lines between the includes and the rest of the code

SafwatHalaby commented:
It took me a while to realize that I'm actually using `PathDestination` for two distinct purposes. I split that into two variables and that solved most issues.

SafwatHalaby commented:
Do not merge yet, even if CI succeeds.

SafwatHalaby commented:
Rebased.
Looking good, but I'm still not satisfied with idle mob behavior. I'll need to rethink a bit.

SafwatHalaby commented:
The last commit fixes idle mob issues.
I changed the `PathFinder.h` header a bit.
I Added a `dontCare` bool. If set to true, it means the monster doesn't care where it would end up going, and the PathFinder  may change the monster's final destination.

SafwatHalaby commented:
Also, I used a pointer because it explicitly indicates that the Pathfinder may modify the finalDestination. WIth references, I found myself putting comments all over the place to prevent confusion.

SafwatHalaby commented:
lgtm

SafwatHalaby commented on src/Mobs/PathFinder.h:
```diff
@@ -0,0 +1,96 @@
+	@param a_Destination The position the mob would like to reach. If a_ExactPath is true, the PathFinder may modify this.
+	@param a_OutputWaypoint An output parameter: The next waypoint to go to.
+	@param a_DontCare If true, the mob doesn't care where to go, and the Pathfinder may modify a_Destination.
+	This should usually be true. An exception is a wandering idle mob which doesn't care about its final destination.
```
This should usually be **false**.

Seadragon91 commented:
There are still a few TODOs. Are they treated in a other pull request, someday?

SafwatHalaby commented:
@Seadragon91 Which TODOs are you referring to?

SafwatHalaby commented:
@Seadragon91 
Monster give up timer - No longer relevant
monster.h TODO - Done
goto TODO - Will do later

lgtm.
